File: tw-gradcs-linear-algebra/04-orthogonality-and-least-squares/02-projections/c/projection.c
/**
 * 投影 (Projections)
 *
 * 本程式示範：
 * 1. 投影到直線
 * 2. 投影矩陣及其性質
 * 3. 誤差向量的正交性驗證
 *
 * 編譯：gcc -std=c99 -O2 projection.c -o projection -lm
 * 執行：./projection
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>

#define EPSILON 1e-10
#define MAX_DIM 10

// ========================================
// 輔助函數
// ========================================

void print_separator(const char* title) {
    printf("\n");
    for (int i = 0; i < 60; i++) printf("=");
    printf("\n%s\n", title);
    for (int i = 0; i < 60; i++) printf("=");
    printf("\n");
}

void print_vector(const char* name, const double* v, int n) {
    printf("%s = [", name);
    for (int i = 0; i < n; i++) {
        printf("%.4f", v[i]);
        if (i < n - 1) printf(", ");
    }
    printf("]\n");
}

void print_matrix(const char* name, double M[MAX_DIM][MAX_DIM], int rows, int cols) {
    printf("%s =\n", name);
    for (int i = 0; i < rows; i++) {
        printf("  [");
        for (int j = 0; j < cols; j++) {
            printf("%8.4f", M[i][j]);
            if (j < cols - 1) printf(", ");
        }
        printf("]\n");
    }
}

// ========================================
// 基本運算
// ========================================

double dot_product(const double* x, const double* y, int n) {
    double result = 0.0;
    for (int i = 0; i < n; i++) {
        result += x[i] * y[i];
    }
    return result;
}

double vector_norm(const double* x, int n) {
    return sqrt(dot_product(x, x, n));
}

void scalar_multiply(double c, const double* x, double* result, int n) {
    for (int i = 0; i < n; i++) {
        result[i] = c * x[i];
    }
}

void vector_subtract(const double* x, const double* y, double* result, int n) {
    for (int i = 0; i < n; i++) {
        result[i] = x[i] - y[i];
    }
}

void outer_product(const double* x, const double* y, double result[MAX_DIM][MAX_DIM], int m, int n) {
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            result[i][j] = x[i] * y[j];
        }
    }
}

void matrix_scalar_multiply(double c, double A[MAX_DIM][MAX_DIM], double result[MAX_DIM][MAX_DIM], int m, int n) {
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            result[i][j] = c * A[i][j];
        }
    }
}

void matrix_vector_multiply(double A[MAX_DIM][MAX_DIM], const double* x, double* result, int m, int n) {
    for (int i = 0; i < m; i++) {
        result[i] = 0.0;
        for (int j = 0; j < n; j++) {
            result[i] += A[i][j] * x[j];
        }
    }
}

void matrix_multiply(double A[MAX_DIM][MAX_DIM], double B[MAX_DIM][MAX_DIM],
                     double result[MAX_DIM][MAX_DIM], int m, int k, int n) {
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            result[i][j] = 0.0;
            for (int p = 0; p < k; p++) {
                result[i][j] += A[i][p] * B[p][j];
            }
        }
    }
}

// ========================================
// 投影函數
// ========================================

typedef struct {
    double x_hat;
    double projection[MAX_DIM];
    double error[MAX_DIM];
    double error_norm;
} ProjectionResult;

/**
 * 投影到直線
 * p = (aᵀb / aᵀa) * a
 */
ProjectionResult project_onto_line(const double* b, const double* a, int n) {
    ProjectionResult result;

    double aTb = dot_product(a, b, n);
    double aTa = dot_product(a, a, n);

    result.x_hat = aTb / aTa;
    scalar_multiply(result.x_hat, a, result.projection, n);
    vector_subtract(b, result.projection, result.error, n);
    result.error_norm = vector_norm(result.error, n);

    return result;
}

/**
 * 投影到直線的投影矩陣
 * P = aaᵀ / (aᵀa)
 */
void projection_matrix_line(const double* a, double P[MAX_DIM][MAX_DIM], int n) {
    double aTa = dot_product(a, a, n);
    double aaT[MAX_DIM][MAX_DIM];

    outer_product(a, a, aaT, n, n);
    matrix_scalar_multiply(1.0 / aTa, aaT, P, n, n);
}

/**
 * 驗證投影矩陣的性質
 */
void verify_projection_matrix(double P[MAX_DIM][MAX_DIM], int n, const char* name) {
    printf("\n驗證 %s 的性質：\n", name);

    // 對稱性
    bool is_symmetric = true;
    for (int i = 0; i < n && is_symmetric; i++) {
        for (int j = 0; j < n && is_symmetric; j++) {
            if (fabs(P[i][j] - P[j][i]) > EPSILON) {
                is_symmetric = false;
            }
        }
    }
    printf("  對稱性 (%sᵀ = %s)：%s\n", name, name, is_symmetric ? "true" : "false");

    // 冪等性
    double P2[MAX_DIM][MAX_DIM];
    matrix_multiply(P, P, P2, n, n, n);

    bool is_idempotent = true;
    for (int i = 0; i < n && is_idempotent; i++) {
        for (int j = 0; j < n && is_idempotent; j++) {
            if (fabs(P[i][j] - P2[i][j]) > EPSILON) {
                is_idempotent = false;
            }
        }
    }
    printf("  冪等性 (%s² = %s)：%s\n", name, name, is_idempotent ? "true" : "false");
}

// ========================================
// 主程式
// ========================================

int main() {
    print_separator("投影示範 (C)\nProjection Demo");

    // 1. 投影到直線
    print_separator("1. 投影到直線");

    double a[] = {1.0, 1.0};
    double b[] = {2.0, 0.0};

    print_vector("方向 a", a, 2);
    print_vector("向量 b", b, 2);

    ProjectionResult result = project_onto_line(b, a, 2);

    printf("\n投影係數 x̂ = (aᵀb)/(aᵀa) = %.4f\n", result.x_hat);
    print_vector("投影 p = x̂a", result.projection, 2);
    print_vector("誤差 e = b - p", result.error, 2);

    // 驗證正交性
    double e_dot_a = dot_product(result.error, a, 2);
    printf("\n驗證 e ⊥ a：e · a = %.6f\n", e_dot_a);
    printf("正交？ %s\n", fabs(e_dot_a) < EPSILON ? "true" : "false");

    // 2. 投影矩陣
    print_separator("2. 投影矩陣（到直線）");

    double a2[] = {1.0, 2.0};
    print_vector("方向 a", a2, 2);

    double P[MAX_DIM][MAX_DIM];
    projection_matrix_line(a2, P, 2);
    print_matrix("\n投影矩陣 P = aaᵀ/(aᵀa)", P, 2, 2);

    verify_projection_matrix(P, 2, "P");

    // 用投影矩陣計算投影
    double b2[] = {3.0, 4.0};
    print_vector("\n向量 b", b2, 2);

    double p[MAX_DIM];
    matrix_vector_multiply(P, b2, p, 2, 2);
    print_vector("投影 p = Pb", p, 2);

    // 3. 多個向量的投影
    print_separator("3. 批次投影");

    double vectors[][2] = {{1.0, 0.0}, {0.0, 1.0}, {2.0, 2.0}, {3.0, -1.0}};
    int num_vectors = 4;

    printf("方向 a = [1, 2]\n");
    printf("\n各向量投影結果：\n");

    for (int i = 0; i < num_vectors; i++) {
        ProjectionResult proj = project_onto_line(vectors[i], a2, 2);
        printf("  [%.1f, %.1f] -> [%.4f, %.4f]\n",
               vectors[i][0], vectors[i][1],
               proj.projection[0], proj.projection[1]);
    }

    // 總結
    print_separator("總結");
    printf("\n");
    printf("投影公式：\n\n");
    printf("1. 投影到直線：\n");
    printf("   p = (aᵀb / aᵀa) a\n");
    printf("   P = aaᵀ / (aᵀa)\n\n");
    printf("2. 投影到子空間：\n");
    printf("   p = A(AᵀA)⁻¹Aᵀb\n");
    printf("   P = A(AᵀA)⁻¹Aᵀ\n\n");
    printf("3. 投影矩陣性質：\n");
    printf("   Pᵀ = P（對稱）\n");
    printf("   P² = P（冪等）\n\n");

    for (int i = 0; i < 60; i++) printf("=");
    printf("\n示範完成！\n");
    for (int i = 0; i < 60; i++) printf("=");
    printf("\n");

    return 0;
}

File: tw-gradcs-linear-algebra/04-orthogonality-and-least-squares/02-projections/javascript/projection.js
/**
 * 投影 (Projections)
 *
 * 本程式示範：
 * 1. 投影到直線
 * 2. 投影矩陣及其性質
 * 3. 誤差向量的正交性驗證
 *
 * 執行：node projection.js
 */

const EPSILON = 1e-10;

// ========================================
// 輔助函數
// ========================================

function printSeparator(title) {
    console.log();
    console.log('='.repeat(60));
    console.log(title);
    console.log('='.repeat(60));
}

function formatNumber(n) {
    return n.toFixed(4);
}

function printVector(name, v) {
    const formatted = v.map(x => formatNumber(x)).join(', ');
    console.log(`${name} = [${formatted}]`);
}

function printMatrix(name, M) {
    console.log(`${name} =`);
    for (const row of M) {
        const formatted = row.map(x => formatNumber(x).padStart(8)).join(', ');
        console.log(`  [${formatted}]`);
    }
}

// ========================================
// 基本運算
// ========================================

function dotProduct(x, y) {
    let result = 0;
    for (let i = 0; i < x.length; i++) {
        result += x[i] * y[i];
    }
    return result;
}

function vectorNorm(x) {
    return Math.sqrt(dotProduct(x, x));
}

function scalarMultiply(c, x) {
    return x.map(xi => c * xi);
}

function vectorSubtract(x, y) {
    return x.map((xi, i) => xi - y[i]);
}

function outerProduct(x, y) {
    return x.map(xi => y.map(yj => xi * yj));
}

function matrixScalarMultiply(c, A) {
    return A.map(row => row.map(val => c * val));
}

function matrixVectorMultiply(A, x) {
    return A.map(row => dotProduct(row, x));
}

function matrixMultiply(A, B) {
    const m = A.length, n = B[0].length, k = B.length;
    const result = [];
    for (let i = 0; i < m; i++) {
        result[i] = [];
        for (let j = 0; j < n; j++) {
            result[i][j] = 0;
            for (let p = 0; p < k; p++) {
                result[i][j] += A[i][p] * B[p][j];
            }
        }
    }
    return result;
}

// ========================================
// 投影函數
// ========================================

/**
 * 投影到直線
 * p = (aᵀb / aᵀa) * a
 */
function projectOntoLine(b, a) {
    const aTb = dotProduct(a, b);
    const aTa = dotProduct(a, a);

    const xHat = aTb / aTa;
    const p = scalarMultiply(xHat, a);
    const e = vectorSubtract(b, p);

    return {
        xHat,
        projection: p,
        error: e,
        errorNorm: vectorNorm(e)
    };
}

/**
 * 投影到直線的投影矩陣
 * P = aaᵀ / (aᵀa)
 */
function projectionMatrixLine(a) {
    const aTa = dotProduct(a, a);
    const aaT = outerProduct(a, a);
    return matrixScalarMultiply(1.0 / aTa, aaT);
}

/**
 * 驗證投影矩陣的性質
 */
function verifyProjectionMatrix(P, name = 'P') {
    const n = P.length;

    console.log(`\n驗證 ${name} 的性質：`);

    // 對稱性
    let isSymmetric = true;
    for (let i = 0; i < n && isSymmetric; i++) {
        for (let j = 0; j < n && isSymmetric; j++) {
            if (Math.abs(P[i][j] - P[j][i]) > EPSILON) {
                isSymmetric = false;
            }
        }
    }
    console.log(`  對稱性 (${name}ᵀ = ${name})：${isSymmetric}`);

    // 冪等性
    const P2 = matrixMultiply(P, P);
    let isIdempotent = true;
    for (let i = 0; i < n && isIdempotent; i++) {
        for (let j = 0; j < n && isIdempotent; j++) {
            if (Math.abs(P[i][j] - P2[i][j]) > EPSILON) {
                isIdempotent = false;
            }
        }
    }
    console.log(`  冪等性 (${name}² = ${name})：${isIdempotent}`);
}

// ========================================
// 主程式
// ========================================

function main() {
    printSeparator('投影示範 (JavaScript)\nProjection Demo');

    // 1. 投影到直線
    printSeparator('1. 投影到直線');

    const a = [1, 1];
    const b = [2, 0];

    printVector('方向 a', a);
    printVector('向量 b', b);

    const result = projectOntoLine(b, a);

    console.log(`\n投影係數 x̂ = (aᵀb)/(aᵀa) = ${formatNumber(result.xHat)}`);
    printVector('投影 p = x̂a', result.projection);
    printVector('誤差 e = b - p', result.error);

    // 驗證正交性
    const eDotA = dotProduct(result.error, a);
    console.log(`\n驗證 e ⊥ a：e · a = ${eDotA.toFixed(6)}`);
    console.log(`正交？ ${Math.abs(eDotA) < EPSILON}`);

    // 2. 投影矩陣
    printSeparator('2. 投影矩陣（到直線）');

    const a2 = [1, 2];
    printVector('方向 a', a2);

    const P = projectionMatrixLine(a2);
    printMatrix('\n投影矩陣 P = aaᵀ/(aᵀa)', P);

    verifyProjectionMatrix(P);

    // 用投影矩陣計算投影
    const b2 = [3, 4];
    printVector('\n向量 b', b2);

    const p = matrixVectorMultiply(P, b2);
    printVector('投影 p = Pb', p);

    // 3. 多個向量的投影
    printSeparator('3. 批次投影');

    const vectors = [[1, 0], [0, 1], [2, 2], [3, -1]];

    console.log('方向 a = [1, 2]');
    console.log('\n各向量投影結果：');

    for (const v of vectors) {
        const proj = projectOntoLine(v, a2);
        console.log(`  [${v[0]}, ${v[1]}] -> [${formatNumber(proj.projection[0])}, ${formatNumber(proj.projection[1])}]`);
    }

    // 總結
    printSeparator('總結');
    console.log(`
投影公式：

1. 投影到直線：
   p = (aᵀb / aᵀa) a
   P = aaᵀ / (aᵀa)

2. 投影到子空間：
   p = A(AᵀA)⁻¹Aᵀb
   P = A(AᵀA)⁻¹Aᵀ

3. 投影矩陣性質：
   Pᵀ = P（對稱）
   P² = P（冪等）
`);

    console.log('='.repeat(60));
    console.log('示範完成！');
    console.log('='.repeat(60));
}

main();
